ktor-server/ktor-server-core/src/io/ktor/features/ContentNegotiation.kt

```diff
@@ -65,7 +65,13 @@ class ContentNegotiation(val registrations: List<ConverterRegistration>) {
             pipeline.sendPipeline.intercept(ApplicationSendPipeline.Render) {
                 if (subject is OutgoingContent) return@intercept
 
-                val suitableConverters = call.request.acceptItems().mapNotNull { (contentType, _) ->
+                var acceptContentTypes = call.request.acceptItems().map { it.value }
+                // From https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
+                // "If no Accept header field is present, then it is assumed that the client accepts all media types."
+                if(acceptContentTypes.isEmpty()) {
```

Comment:
I think it would be better to just use `firstOrNull` on registrations instead of going through all the hoops to match `*/*`

Reply:
Makes sense. See new commit.

Reply:
Note that we still need to run every suitableRegistration until we've found a corresponding converter that can handle the value.

Reply:
Hmm... nevermind, I think I need to look at the logic more carefully. I'll get back.

Reply:
Ok, I had another look. Which do you prefer from the following two alternatives?

```kotlin
val acceptItems = call.request.acceptItems()
val suitableConverters = if (acceptItems.isEmpty()) {
    feature.registrations.take(1)
} else {
    acceptItems.mapNotNull { (contentType, _) ->
        feature.registrations.firstOrNull {
            it.contentType.match(contentType)
        }
    }
}

// Pick the first one that can convert the subject successfully
val converted = suitableConverters.mapUntilNotNull {
    it.converter.convertForSend(this, it.contentType, subject)
}
```

```kotlin
val acceptItems = call.request.acceptItems()
val suitableConverters = if (acceptItems.isEmpty()) {
    feature.registrations
} else {
    acceptItems.flatMap { (contentType, _) ->  // flatMap here
        feature.registrations.filter {  // filter here
            it.contentType.match(contentType)
        }
    }
}

// Pick the first one that can convert the subject successfully
val converted = suitableConverters.mapUntilNotNull {
    it.converter.convertForSend(this, it.contentType, subject)
}
```

The second alternative seems more general to me. We will try all registered converters that matches the Accept field, and pick the first one that succeeds.

In the first case, we will only try to convert with the *first* converter (for each Accept-statement).

What do you think?